File: arlib/monabs/cores/unary_check.py
"""
Unary satisfiability checks for lists of constraints.

The helpers here answer, for each constraint under a shared precondition:
- 1: constraint is satisfiable
- 0: constraint is unsatisfiable
- 2: solver returned unknown

Variants cover basic, incremental (push/pop), and cache-aware modes that
re-use models to mark additional constraints as satisfiable when possible.
"""
from typing import List
import z3


def unary_check(precond: z3.ExprRef, cnt_list: List[z3.ExprRef]) -> List:
    """Check each constraint independently under the precondition (fresh solver per check)."""
    results = [None] * len(cnt_list)

    for i, cnt in enumerate(cnt_list):
        solver = z3.Solver()
        solver.add(precond)  # Add the precondition
        solver.add(cnt)  # Add the current constraint
        res = solver.check()
        if res == z3.sat:
            results[i] = 1
        elif res == z3.unsat:
            results[i] = 0
        else:
            results[i] = 2

    return results


def unary_check_incremental(precond: z3.ExprRef, cnt_list: List[z3.ExprRef]) -> List:
    """Check each constraint with a shared solver using push/pop for efficiency."""
    results = [None] * len(cnt_list)
    solver = z3.Solver()

    solver.add(precond)  # Add the precondition
    for i, cnt in enumerate(cnt_list):
        solver.push()  # Save the current state

        solver.add(cnt)  # Add the current constraint
        res = solver.check()
        if res == z3.sat:
            results[i] = 1
        elif res == z3.unsat:
            results[i] = 0
        else:
            results[i] = 2

        solver.pop()  # Restore the state

    return results


def unary_check_cached(precond: z3.ExprRef, cnt_list: List[z3.ExprRef]) -> List:
    """Reuse satisfying models to mark other constraints true when implied by the model."""
    results = [None] * len(cnt_list)

    for i, cnt in enumerate(cnt_list):
        if results[i] is not None:
            continue

        solver = z3.Solver()
        solver.add(precond)  # Add the precondition
        solver.add(cnt)  # Add the current constraint
        res = solver.check()
        if res == z3.sat:
            model = solver.model()
            results[i] = 1
            for j, other_cnt in enumerate(cnt_list):
                if results[j] is None and z3.is_true(model.eval(other_cnt, model_completion=True)):
                    results[j] = 1
        elif res == z3.unsat:
            results[i] = 0
        else:
            results[i] = 2

    return results


def unary_check_incremental_cached(precond: z3.ExprRef, cnt_list: List[z3.ExprRef]) -> List:
    """Incremental + caching: share solver state and propagate model truths across constraints."""
    results = [None] * len(cnt_list)
    solver = z3.Solver()

    solver.add(precond)  # Add the precondition

    for i, cnt in enumerate(cnt_list):
        if results[i] is not None:
            continue

        solver.push()  # Save the current state

        solver.add(cnt)  # Add the current constraint
        res = solver.check()
        if res == z3.sat:
            model = solver.model()
            results[i] = 1
            for j, other_cnt in enumerate(cnt_list):
                if results[j] is None and z3.is_true(model.eval(other_cnt, model_completion=True)):
                    results[j] = 1
        elif res == z3.unsat:
            results[i] = 0
        else:
            results[i] = 2

        solver.pop()  # Restore the state

    return results

File: aria/monabs/cpp/unary_check.cpp
// Unary satisfiability helpers.
#include "monabs.hpp"
#include <optional>
#include <vector>
#include <z3++.h>

#include "detail.hpp"

namespace aria::monabs::app {
namespace {

using detail::ensure_same_context;
using detail::all_indices;
using detail::map_check_result;
using detail::model_true;
using detail::propagate_model_sat;

}  // namespace

ResultVector unary_check(const z3::expr& precond, const std::vector<z3::expr>& constraints) {
    ensure_same_context(precond, constraints);
    ResultVector results;
    results.reserve(constraints.size());
    auto& ctx = precond.ctx();
    for (const auto& cnt : constraints) {
        z3::solver solver(ctx);
        solver.add(precond);
        solver.add(cnt);
        results.push_back(map_check_result(solver.check()));
    }
    return results;
}

ResultVector unary_check_incremental(const z3::expr& precond, const std::vector<z3::expr>& constraints) {
    ensure_same_context(precond, constraints);
    ResultVector results;
    results.reserve(constraints.size());
    z3::solver solver(precond.ctx());
    solver.add(precond);
    for (const auto& cnt : constraints) {
        solver.push();
        solver.add(cnt);
        results.push_back(map_check_result(solver.check()));
        solver.pop();
    }
    return results;
}

namespace {
void handle_check_result(z3::check_result res, z3::solver& solver, std::size_t idx,
                         const std::vector<z3::expr>& constraints,
                         std::vector<std::optional<CheckResult>>& results) {
    if (res == z3::sat) {
        const auto model = solver.get_model();
        results[idx] = CheckResult::Sat;
        propagate_model_sat(model, constraints, results, all_indices(constraints.size()));
    } else {
        results[idx] = (res == z3::unsat) ? CheckResult::Unsat : CheckResult::Unknown;
    }
}
}  // namespace

ResultVector unary_check_cached(const z3::expr& precond, const std::vector<z3::expr>& constraints) {
    ensure_same_context(precond, constraints);
    std::vector<std::optional<CheckResult>> results(constraints.size());
    auto& ctx = precond.ctx();
    for (std::size_t i = 0; i < constraints.size(); ++i) {
        if (results[i].has_value()) continue;
        z3::solver solver(ctx);
        solver.add(precond);
        solver.add(constraints[i]);
        handle_check_result(solver.check(), solver, i, constraints, results);
    }
    return detail::finalize_results(results);
}

ResultVector unary_check_incremental_cached(const z3::expr& precond,
                                            const std::vector<z3::expr>& constraints) {
    ensure_same_context(precond, constraints);
    std::vector<std::optional<CheckResult>> results(constraints.size());
    z3::solver solver(precond.ctx());
    solver.add(precond);
    for (std::size_t i = 0; i < constraints.size(); ++i) {
        if (results[i].has_value()) continue;
        solver.push();
        solver.add(constraints[i]);
        handle_check_result(solver.check(), solver, i, constraints, results);
        solver.pop();
    }
    return detail::finalize_results(results);
}

std::string to_string(CheckResult result) {
    constexpr const char* strs[] = {"unsat(0)", "sat(1)", "unknown(2)"};
    return strs[static_cast<int>(result)];
}

}  // namespace aria::monabs::app
